### ISSUE TITLE
Bug: infinite redirect loop with Duo 2FA

### ISSUE THREAD

**Issue by xaque** [marklieberman/foxygestures#257]
When I try to login to my university site, I get into this redirect loop. I figured it out to be something with the foxy gestures extension just by disabling extensions until it worked. I tried using the foxy gestures settings to blacklist the logon page and all the pages it makes any requests to in that loop to no avail. (The blacklisting did stop gestures but didn't stop the bug.)

I think it has something to do with the 2FA that the login uses [(Duo)](https://duo.com/) because that is where the it loops as seen here:
![redirect loop](https://user-images.githubusercontent.com/11743019/42437707-217368ca-831b-11e8-93a9-9fcccf783520.png)

I am using:

- Firefox 61.0.1
- Linux kernel 4.17.4-1-ARCH
- Foxy Gestures 1.1.1

although this has been happening as long as I have been using Foxy. This hasn't happened with any other sites for me, but there aren't any other with which I've used Duo 2FA.

**Comment by marklieberman** [comment]
I believe it has to do with Duo not properly checking if postMessage() messages belong to Duo, or are from another source. In other words its probably lousy code on Duo's side. However, I've never been able to debug it because nobody can provide me with an example page and login to test.

**Comment by xaque** [comment]
I tried out the Duos demos at https://demo.duo.com/ but the redirect loop doesn't happen. If you need something to test with, it looks like anyone can create a logon from my university web page https://cas.byu.edu/cas/login by just clicking on "create a netid". Then enrolling in 2fa from https://duo.byu.edu/

**Comment by marklieberman** [comment]
Here is what I know from debugging:

1) Duo has a form #duo_form with an action to submit the login.
2) Duo adds an iframe #duo_iframe to the page that does the 2FA stuff.
3) Any message with an origin of the #duo_frame gets handled by this code:

```js
.receiveMessage(function (K) {
  K = K.data + ':' + Duo._app_sig;
  K = D('<input type="hidden">').attr('name', Duo._post_argument).val(K);
  var L = D('#duo_form');
  if (!L.length) {
    L = D('<form>');
    L.insertAfter(t)
  }
  L.attr('method', 'POST');
  L.attr('action', Duo._post_action);
  L.append(K);
  L.submit()
}, 'https://' + Duo._host)
```

4) That code does not do any check to ensure the message (K.data) is the expected message from Duo.
5) When Foxy Gestures loads in a frame, it uses postMessage() to notify the parent frame that it is ready and establish a hierarchy.
6) L.submit() submits #duo_form which causes the page reload because K.data isn't the correct login stuff.

Next step is looking into a solution.

**Comment by marklieberman** [comment]
@xaque I've published version 1.2.1 to AMO which resolves the infinite redirect for me. Gestures inside the Duo iframe will not work however, because I've simply excluded the duosecurity.com domain from being injected with content scripts. Please ensure that you've updated to 1.2.1 and then let me know if you are able to login. Thanks

**Comment by xaque** [comment]
It's working flawlessly, thanks for your work on this. Hopefully this helps more people than just me :)
Cheers.
